core/src/main/java/jenkins/util/SystemPropertiesProvider.java

```diff
@@ -0,0 +1,79 @@
+package jenkins.util;
+
+import hudson.util.CopyOnWriteMap;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * Custom Provider extension.
+ * @author Oleg Nenashev
+ * @see SystemProperties
+ * @see ServletContextSystemPropertiesProvider
+ */
+@Restricted(NoExternalUse.class)
+public abstract class SystemPropertiesProvider {
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(SystemPropertiesProvider.class.getName());
+
+    //TODO: This should use the annotation indexer at least.
+    // Ideally it should be a part of the Jenkins singleton in order to play nicely with {@code JenkinsRule}.
+    private static final CopyOnWriteMap<String, SystemPropertiesProvider> providers = new CopyOnWriteMap.Hash<>();
```

Comment:
the map symatics are not really unused - you could remove using the actual provider (object) to remove and use a standard CopyOnWrite collection.

Ontop of this you get different (random) iteration order (the CopyOnWriteMap makes not guarntees about ordering... so could get hard to reproduce bugs when you sometimes get a system property and other times get a ServletContext parameter...

Reply:
Yeah, probably Map is not required. I just added it to support Removal, but there are more simple and efficient ways for sure

Reply:
Well, this is all restricted, and there is no removal, so KISS.